keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,365 @@
+     - If Kubelet determines there isn't enough room, it simply retries the Pod
+       resize at a later time.
+1. Scheduler uses cached Pod's Spec.Containers[i].ResourcesAllocated to compute
+   resources available on the Node while a Pod resize may be in progress.
+   * If a new Pod is assigned to that Node in parallel, it can temporarily
+     result in actual sum of Pod resources for the Node exceeding Node's
+     allocable resources. This is resolved when Kubelet rejects that new Pod
+     during admission due to lack of room.
+   * Once Kubelet that accepted a parallel Pod resize updates that Pod's
+     Spec.Containers[i].ResourcesAllocated, and subsequently the Scheduler
+     updates its cache, accounting will reflect updated Pod resources for
+     future computations and scheduling decisions.
+1. The initiating actor (e.g. VPA) observes the following:
+   * _Case 1_: Pod's Spec.Containers[i].ResourcesAllocated values have changed
+     and matches Spec.Containers[i].Resources.Requests, signifying that desired
+     resize has been accepted, and Pod is being resized. The resize operation
+     is complete when Pod's Status.ContainerStatuses[i].Resources and
+     Spec.Containers[i].Resources match.
+   * _Case 2_: Pod's Spec.Containers[i].ResourcesAllocated remains unchanged,
+     and continues to differ from desired Spec.Containers[i].Resources.Requests.
+     After a certain (user defined) timeout, initiating actor may take alternate
+     action. For example, based on Retry policy, initiating actor may:
+     - Evict the Pod to trigger a replacement Pod with new desired resources,
+     - Do nothing and let Kubelet back off and later retry the in-place resize.
+
+#### Container resource limit update ordering
+
+When in-place resize is requested for multiple Containers in a Pod, Kubelet
+updates resource limit for the Pod and its Containers in the following manner:
+  1. If resource resizing results in net-increase of a resource type (CPU or
+     Memory), Kubelet first updates Pod-level cgroup limit for the resource
+     type, and then updates the Container resource limit.
+  1. If resource resizing results in net-decrease of a resource type, Kubelet
+     first updates the Container resource limit, and then updates Pod-level
+     cgroup limit.
+  1. If resource update results in no net change of a resource type, only the
+     Container resource limits are updated.
+
+In all the above cases, Kubelet applies Container resource limit decreases
+before applying limit increases.
+
+#### Notes
+
+* If CPU Manager policy for a Node is set to 'static', then only integral
+  values of CPU resize are allowed. If non-integral CPU resize is requested
+  for a Node with 'static' CPU Manager policy, that resize is rejected, and
+  an error message is logged to the event stream.
+* To avoid races and possible gamification, all components will use Pod's
+  Spec.Containers[i].ResourcesAllocated when computing resources used by Pods.
```

Comment:
How is this going to work with ResourceQuota? Is that Admission Controller supposed to watch (& veto) a change made by kubelet?
I'd imagine that using max(Resources, ResourcesAllocated) prevents gamification better since the user cannot artificially lower the resource usage nor abuse some race condition to temporarily used lower values.

Reply:
True. I highlighted this point in last-week's sig-node meeting - the case where Kubelet whose Pod is being sized down may be offline, and until that comes back up there is potential that actual cluster resources used by a NS might exceed the quota limit.

With introduction or ResourcesAllocated in PodSpec, the [amount of lag](https://github.com/kubernetes/enhancements/pull/686/#discussion_r324809149) for sizing down should be very minimal (the watch latency at Kubelet), and for Limits, we are always expected to converge towards desired limits. So the consensus was to go with the the simplicity of just using PodSpec.Resources, and if we saw any major problems this is a small change (few lines of code) to go with max.

Reply:
Resolving this issue. If we discover problems with just using Resources instead of max, then we can easily change the code. I'm tracking it in my list of issues so we test for it explicitly.